### PR TITLE
Cancellable reads; EOF on "read" sets var to ""

### PR DESCRIPTION
**Pull request by theclapp**
- Make reading cancellable. Not all input from stdin, just that done directly by the shell (i.e. the "read" builtin). Exec'ed programs still read directly from stdin's os.File and are not cancellable.
- If you press ^D (EOF) when reading into a shell variable, set the variable to "". This is consistent with bash & zsh.

Fixes https://github.com/mvdan/sh/issues/856.
See also https://github.com/mvdan/sh/pull/857 for my first try at this.

### CODE REVIEW

**Review by mvdan — COMMENTED** [review]
I assume this is one of those that is hard to test properly, but I'd still like us to try, because the logic is non-trivial and feels fragile.

For example, with https://pkg.go.dev/os#Pipe to create a stdin file, and cancelling the runner context to mimic ^D, we could test a few edge cases with the `read` builtin:
* only newline (empty input)
* only ^D (empty input)
* some string and newline (non-empty input)
* some string and ^D (non-empty input)

**Comment by mvdan** [comment]
Overall nothing against merging this feature though - it is one added dependency, but it seems like a light one, and I don't think we can possibly implement cancelling reads any way other than OS-specific code.

File: interp/builtin.go [hunk 1]
```diff
@@ -606,6 +604,10 @@ func (r *Runner) builtinCode(ctx context.Context, pos syntax.Pos, name string, a
 			r.setVarString(name, val)
 		}
 
+		if err != nil {
+			return 1
+		}
```
[the remark this answers is not in the document]
Comment (theclapp) [reply interp/builtin.go:609]: > can you document why we check the error after using the returned value from readLine now?

In retrospect, it's not really clear, but I mentioned it in the PR comment: so that if you press ^D (EOF) when reading into a shell variable, it sets the variable to "".

I'll be happy to add a comment to the actual code, of course. This is just so you know.

**Comment by theclapp** [comment]
> I assume this is one of those that is hard to test properly, but I'd still like us to try, because the logic is non-trivial and feels fragile.

Sure, no worries.

**Comment by theclapp** [comment]
> * only newline (empty input)
> * only ^D (empty input)
> * some string and newline (non-empty input)
> * some string and ^D (non-empty input)

It turns out the current code already works with the first, third, and fourth. I've added a test case with the second.

None of these cases actually test having a read in-progress and cancelling its context, so I wrote a separate test for that. It turns out it wasn't that difficult.

File: interp/interp_test.go [hunk 2]
```diff
@@ -2982,6 +2982,27 @@ done <<< 2`,
...
+	// // some string and newline
+	// {
+	// 	"a=c; echo x | read a; echo -n $a",
+	// 	"x",
+	// },
+	// // some string and EOF
+	// {
+	// 	"a=d; echo -n y | read a; echo -n $a",
+	// 	"y",
+	// },
```
Comment (mvdan) [inline interp/interp_test.go:3005]: did you mean to leave most of these commented out?

File: interp/interp_test.go [hunk 3]
```diff
@@ -3943,6 +3964,49 @@ func TestRunnerContext(t *testing.T) {
...
+	} else {
+		siR, siW, err := os.Pipe()
+		if err != nil {
+			t.Fatalf("Error calling os.Pipe: %v", err)
+		}
+		defer func() {
+			siW.Close()
+			siR.Close()
+		}()
+		r, _ = interp.New(interp.StdIO(siR, nil, nil))
```
Comment (mvdan) [inline interp/interp_test.go:3990]: nit: move interp.New and StdIO outside the if/else, you can just have each branch set the stdin file var

File: interp/builtin.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mvdan) [reply interp/builtin.go:610]: LGTM

**Review by mvdan — COMMENTED** [review]
FYI CI appears broken

File: interp/interp_test.go (hunk 2 above)
Comment (theclapp) [reply interp/interp_test.go:3005]: Yes. Sort of. I mentioned in the commit:

- Add some test cases. Daniel suggested four, three of which already
  pass without this PR. I left them in but commented-out, in case
  somebody wants to look at them.

Should uncomment them or remove them entirely? Wasn't sure how you'd feel about that either way.

Comment (mvdan) [reply interp/interp_test.go:3005]: Sorry, maybe I'm being slow as it's a bit late. If these test cases already behaved the way we want in master, why not enable them as regression tests, even if they're not directly related to this fix?

Comment (theclapp) [reply interp/interp_test.go:3005]: Sure, I'm fine with that.

**Comment by theclapp** [comment]
As noted in the commit: TestCancelreader failing intermittently under Linux (on a Debian VM), even though the read context is definitely cancelled. Not sure what's going on. Committing what I've got, for posterity.

**Comment by theclapp** [comment]
> TestCancelreader failing intermittently under Linux

At a guess, this is because of, or at least related to, the data race reported in the CI tests. Will look more tomorrow.

**Comment by theclapp** [comment]
@ghostsquad Wanna take a look?

**Review by mvdan — APPROVED** [review]
SGTM, thanks! If anyone else has thoughts or a follow-up review, I'm happy to get more PRs after this lands.

**Comment by mvdan** [comment]
@theclapp do you want to squash your commits into one and write a single commit message? I could squash myself, only keeping the original commit message, but I suspect you might want to keep some of the text from the other commit messages, e.g. about the test.

**Comment by theclapp** [comment]
Squashed 'em all.

**Review by mvdan — APPROVED** [review]
